Update to nixos-23.05

nh2 opened the pull request:
Work on ZuriHac 2023 with @jonathanlking:

* Closes #111.
* Closes #115.

cdepillabout committed:
Bump nixpkgs.nix to pin to nixos-21.11 commit 573095944e7.

cdepillabout committed:
Bump nixpkgs submodule to be on 21.11.

cdepillabout committed:
Make sure the nixpkgs submodule is looking at NixOS/nixpkgs.

cdepillabout committed:
Get the basic example working with 21.11.

cdepillabout committed:
Add more ghc to cabal version mappings.

cdepillabout committed:
Bump to ghc-8.10.7. Nixpkgs is on LTS-18.16.

cdepillabout committed:
postgresl's tests fail when building with musl, so disable postgrest from the set of working Haskell packages

cdepillabout committed:
Apparently there is a curlMinimal top-level derivation now, which also needs to be statified.

cdepillabout committed:
brotli support was added to curl and it needs to be explicitly disabled because it pulls in krb5

cdepillabout committed:
Make sure static archives symlinks are created for openblas.

cdepillabout committed:
Make sure blas-ffi uses openblasCompat.

cdepillabout committed:
Disable stackage executable magico because of gfortran linking error.

cdepillabout committed:
Disable stackage packages that depend on postgres.

cdepillabout committed:
Make sure libdevil has libdeflate as buildInput.

cdepillabout committed:
Drop to postgresql-11 and re-enable all Haskell packages that use postgres.

robx committed:
new ncurses

nh2 committed:
Update to nixos-23.05

nh2 committed:
Fix `approachPkgs` being the wrong value.

It picked the one from the line below it due to `let rec` instead
of the one from the top-level arguments.

nh2 committed:
Add comment on curl brotli disabling

nh2 commented:
Trying

    nix-build --no-out-link survey/default.nix -A working.hello
fails with linker errors regarding libelf.
Investigating the linker invocation (revealed by passing `-v` to `ghc`), we see `... -lelf -ldw ...`.
That's the wrong way around.
Inverting them, we see linker errors regaring symbols from ZSTD and other compression libraries that `libdw` uses.
Adding:

    -L/nix/store/lh5mwy1f0ncakvahryrs3qfxd9pgp81c-zstd-static-x86_64-unknown-linux-musl-1.5.5/lib -lzstd
    -L/nix/store/xyc3qbpkamyjif6aryrdkfwybsax8fqr-zlib-static-x86_64-unknown-linux-musl-1.2.13/lib -lz
    -L/nix/store/vf10zcfikzhxa8lzbr3wy7pfvzyzhr3r-bzip2-static-x86_64-unknown-linux-musl-1.0.8/lib -lbz2
    -L/nix/store/adas6a61mndhcb3sflxgrnpykhdq7acn-xz-static-x86_64-unknown-linux-musl-5.4.3/lib -llzma

Then we get

    ld: /nix/store/13gccvgy3br4jdacyc7ir7h37231d7ib-elfutils-0.189/lib/libdw.a(libdw_alloc.o): in function `__libdw_oom':
    (.text+0x2c0): undefined reference to `error'
    
Unclear what `error` function it is referring to.

* That error looks a bit like what this patch addresses:
  https://github.com/NixOS/nixpkgs/blob/85bcb95aa83be667e562e781e9d186c57a07d757/pkgs/development/tools/misc/elfutils/default.nix#L40
* Scrolling further down, we find
  https://github.com/NixOS/nixpkgs/blob/85bcb95aa83be667e562e781e9d186c57a07d757/pkgs/development/tools/misc/elfutils/default.nix#L95-L96
  `broken = stdenv.hostPlatform.isStatic;` with comment:
  https://lists.fedorahosted.org/pipermail/elfutils-devel/2014-November/004223.html
  This says that `libdwarf` requires `dlopen()` which musl does not support
So instead we just give up on this and set `enableDwarf = false` for our musl GHC build.

nh2 commented:
Debugging why it still has the elf-related linker error after disabling DWARF:

```sh
nix-build --no-out-link --expr '(import ./survey/default.nix {}).working.hello.overrideAttrs (old: { configureFlags = (old.configureFlags or []) ++ ["--ghc-options=-v"]; })'
```

nh2 committed:
Disable GHC dwarf support

nh2 committed:
Add `useArchiveFilesForTemplateHaskell` option.

The motivation is that @aherrmann and @jonathanlking want to use Bazel
and found that it might it easier to use `.a` files in TH than to try
to provide `.so` AND `.a` files of all dependencies.

nh2 commented:
> Debugging why it still has the elf-related linker error after disabling DWARF:

We overrode the GHC incorrectly, overriding only `ghc` but not `buildHaskellPackages`.

nh2 commented:
@jonathanlking, @chpatrick and me added support for `useArchiveFilesForTemplateHaskell` and basic support for `ghc962`.

`ghc962` can build `hello`:

```
nix-build --expr '(import ./survey/default.nix { compiler = "ghc962"; useArchiveFilesForTemplateHaskell = true; }).working.hello'
```

but not much more because the Stackage in nixpkgs is targeting GHC 9.2.
But that's fine because Bazel users likely don't need Haskell packages building in nixpkgs, as they build those themselves.

nh2 committed:
Fix `workingStackageExecutables`'s YAML loading.

See https://stackoverflow.com/questions/69564817/typeerror-load-missing-1-required-positional-argument-loader-in-google-col/69581453#69581453

aherrmann commented on survey/default.nix:
```diff
@@ -1497,10 +1547,39 @@ let
+        then ghcPackage # GHC < 9.6, no Hadrian
+        else ghcPackage.override { enableDwarf = false; }
+    )
+    (ghcPackage:
+      ghcPackage.override { enableRelocatedStaticLibs = useArchiveFilesForTemplateHaskell; }
+    )
```
For the rules_haskell use-case this requires 
```suggestion
    (ghcPackage:
      ghcPackage.override {
        enableRelocatedStaticLibs = useArchiveFilesForTemplateHaskell;
        enableShared = !useArchiveFilesForTemplateHaskell;
       }
    )
```
That is because we require a static runtime to load static Haskell libraries for template Haskell.
Otherwise the build fails with errors like these (tested [here](https://github.com/tweag/skyscope/pull/94)):
```
/nix/store/fhr7av2b52a8dz8lii59bvydgx0vw5ap-binutils-2.39/bin/ld: cannot find -lHScolour-2.3.6-ghc9.0.2: No such file or directory
collect2: error: ld returned 1 exit status
`cc_wrapper-python' failed in phase `Linker'. (Exit code: 1)
```

nh2 committed:
Add nixpkgs patch to add .a files to bzip2 and zstd

nh2 committed:
Make `.working` pass

nh2 committed:
Remove old compilers:

Those no longer packaged by nixpkgs, and 8.8, so that we can assume >= 8.10.

nh2 committed:
Remove workarounds no longer necessary with nixos-23.05

nh2 commented:
@jonathanlking @aherrmann

I pushed some new fixes.

Now only 9 packages fail for `workingStackageExecutables`.

nh2 committed:
Fix arbtt needing libxcb

nh2 committed:
Fix more packages. `workingStackageExecutables` builds on top of nixos-23.05.

Only minimal patches are put on top of the `nixos-23.05` branch to
support more of Stackage. All have nixpkgs PRs.

    nix-build --no-link survey/default.nix -A workingStackageExecutables | wc -l
    368

nh2 committed:
Implement suggestion for bazel / rules_haskell

From https://github.com/nh2/static-haskell-nix/pull/116#pullrequestreview-1474795596

nh2 commented:
I have fixed a few more of the `workingStackageExecutables` set, and made some more nixpkgs PRs (see abvoe).

368 Stackage executables build statically:

```
% nix-build --no-link survey/default.nix -A workingStackageExecutables | wc -l
368
```

The 28 that do not build are:

```
Agda
Allure
csg
cuda
debug
diagrams-builder
diagrams-cairo
gloss-examples
gtk-sni-tray
gtk3
H
hackage-cli
hamilton
hquantlib
ihaskell
LambdaHack
language-puppet
learn-physics
LPFP
magico
monomer
nanovg
odbc
pango
qchas
rhine-gloss
soxlib
termonad
```

Compare this to [previous numbers](https://github.com/nh2/static-haskell-nix/issues/4#issuecomment-406838083).

---

Merging.
